src/java/org/apache/cassandra/db/CounterMutation.java

```diff
@@ -357,38 +361,43 @@ private PartitionUpdate processModifications(PartitionUpdate changes,
         updateWithCurrentValuesFromCFS(marks, cfs, systemClockMicros, counterId);
 
         // What's remain is new counters
-        for (PartitionUpdate.CounterMark mark : marks)
+        for (Pair<PartitionUpdate.CounterMark, CounterCacheKey> mark : marks)
             updateWithCurrentValue(mark, ClockAndCount.BLANK, cfs, systemClockMicros, counterId);
 
         return changes;
     }
 
-    private void updateWithCurrentValue(PartitionUpdate.CounterMark mark,
+    private CounterCacheKey cacheKeyForMark(ColumnFamilyStore cfs, PartitionUpdate.CounterMark mark)
+    {
+        return CounterCacheKey.create(cfs.metadata(), key().getKey(), mark.clustering(), mark.column(), mark.path());
```

Comment:
is it safe to cache the key? I.e. is there a possibility that the buffer comes from a shared pool and may get recycled? Should we duplicate defensively?

Reply:
TLDR: there's no issue

1. if there were an issue, it wouldn't be a regression. The change doesn't impact the lifetime of `CounterCacheKey` (which lived & lives beyond single request handling)
2. having to deal with the issue you mention at this level of abstraction would mean the API is broken
3. the whole problem that we're solving here is copying too much. So yeah, the underlying buffers are already being cloned, and this happens at a lower level
4. It's a pity Java lacks concepts like ownership, non-copyability, immutability, move semantics etc, but that's how it is. Yet again, a case in which we've got to consider issues unknown elsewhere :shrug: